lib/backbeat/Metrics.js

```diff
@@ -208,33 +209,43 @@ class Metrics {
             const now = new Date();
             const timeSinceStart = Math.floor(
                 (now - this._internalStart) / 1000);
+            // Seconds up to a max of EXPIRY seconds
+            const timeDisplay = timeSinceStart < EXPIRY ?
+                timeSinceStart : EXPIRY;
+            const numOfIntervals = Math.ceil(timeDisplay / INTERVAL);
 
             const [opsThroughput, bytesThroughput] = res.map(r => {
-                let total = r.requests.slice(0, 3).reduce(
+                let total = r.requests.slice(0, numOfIntervals).reduce(
                     (acc, i) => acc + i, 0);
 
-                // last interval timestamp
-                const lastInterval =
-                    this._statsClient._normalizeTimestamp(new Date(now));
-                // in seconds
-                const diff = Math.floor((now - lastInterval) / 1000);
+                // if timeDisplay !== EXPIRY, use internal timer and do not
+                // include the extra 4th interval
+                if (timeDisplay === EXPIRY) {
+                    // all intervals apply, including 4th interval
+                    const lastInterval =
+                        this._statsClient._normalizeTimestamp(new Date(now));
+                    // in seconds
+                    const diff = Math.floor((now - lastInterval) / 1000);
 
-                if (timeSinceStart >= EXPIRY) {
-                    // Get average for last interval depending on time surpassed
-                    // so far for newest interval
-                    total += ((INTERVAL - diff) / INTERVAL) * r.requests[3];
+                    if (timeSinceStart >= EXPIRY) {
+                        // Get average for last interval depending on time
+                        // surpassed so far for newest interval
+                        total += ((INTERVAL - diff) / INTERVAL) * r.requests[3];
```

Comment:
The array from `r.requests` is sorted by newest to oldest interval data, so the 4th interval should be the oldest to-be-expired data, so I think the use of `(INTERVAL - diff) / INTERVAL` is accurate.
For the first interval, we collect any data that exists in the incomplete interval. I can weight the 1st interval, but I thought it'd be nice to give the full data even though it might go slightly above the 1 interval total weight of the (1st + 4th) interval weight

Reply:
Oh right! Then the current weighting seems correct, considering the latest interval already accounts for a partial amount of time, we remove that much weight from the oldest interval and sum all intervals to get the average, which overall represents a time window of 3 full intervals.

Thanks for clarifying.